### ISSUE TITLE
Jetpack header size in authentication flow

### ISSUE THREAD

**Issue by johngodley** [Automattic/jetpack#7012]
When connecting to a Jetpack site you end up with a very long URL and HTTP header that on some servers (Nginx, unsure about Apache) can overflow the maximum header size leading to a 502 server error. This only seems to affect username/password login - WPCOM SSO is not affected.

See https://github.com/Automattic/google-docs-add-on/issues/67 and https://github.com/Automattic/google-docs-add-on/issues/65 for background.

Note that this was tested on Nginx 1.10.0 on Ubuntu 16.04.2 with WordPress 4.7.3, Jetpack 4.8.2, and is using FastCGI (the default setup on Ubuntu).

#### Steps to reproduce the issue

To reproduce the problem you need to authorise a Jetpack site using the WordPress.com authorisation flow. A good example of this is the [Google Docs Add-On](https://apps.wordpress.com/google-docs/).

1. Press on WPCOM Connect button and get taken to the `public-api.wordpress.com` oauth2 authorise page:

`https://public-api.wordpress.com/oauth2/authorize`

URL params:

```php
Array
(
    [client_id] => 12345
    [response_type] => code
    [redirect_uri] => [ORIGINAL REDIRECT URI]
    [state] => [355 byte state value]
)
```

2. Select a Jetpack site from the site list and press approve.
3. Get redirected to the Jetpack site:

`http://jetpacksite.thing/wp-login.php`

```php
Array
(
    [action] => jetpack_json_api_authorization
    [redirect_to] => [REDIRECT URI - SEE BELOW]
    [data] => [125 byte value]
    [token] => [TOKEN]
    [timestamp] => 1492675129
    [nonce] => [NONCE]
    [body-hash] =>
    [signature] => [SIGNATURE]
)
```

The `REDIRECT URI` is:

`https://public-api.wordpress.com/oauth2/login/`

```php
Array
(
    [client_id] => 12345
    [response_type] => code
    [redirect_uri] => [ORIGINAL REDIRECT URI]
    [state] => [355 byte state value]
    [action] => oauth2-login
    [redirect_to] => [REDIRECT_TO - see below]
    [blog_id] => 12345
    [_wpnonce] => [NONCE]
)
```

The `REDIRECT_TO` is:

`https://public-api.wordpress.com/oauth2/authorize/`

```php
Array
(
    [client_id] => 12345
    [response_type] => code
    [redirect_uri] => [ORIGINAL REDIRECT URI]
    [state] => [355 byte state value]
    [jetpack-code] =>
    [jetpack-user-id] => 0
    [action] => oauth2-login
)
```

The 355 byte state value is encoded twice in the URL, as is original redirect URI.

4. Login with username/password. **Note that WordPress.com SSO doesn't cause the problem to occur.**

The login page makes a POST request to:

`http://jetpacksite.thing/wp-login.php`

```php
Array
(
    [action] => jetpack_json_api_authorization
    [redirect_to] => [REDIRECT_TO - see below]
    [data] => [125 byte value]
    [token] => [TOKEN]
    [timestamp] => 1492675699
    [nonce] =>[NONCE]
    [body-hash] =>
    [signature] => [SIGNATURE]
)
```

The `REDIRECT_TO` is:

`https://public-api.wordpress.com/oauth2/login/`

```php
Array
(
    [client_id] => 12345
    [response_type] => code
    [redirect_uri] => [ORIGINAL REDIRECT]
    [state] => [355 byte state value]
    [action] => oauth2-login
    [redirect_to] => [REDIRECT_TO - see below]
    [blog_id] => 1234
    [_wpnonce] =>[NONCE]
)
```

The second `REDIRECT_TO` is:

`https://public-api.wordpress.com/oauth2/authorize/`

```php
Array
(
    [client_id] => 12345
    [response_type] => code
    [redirect_uri] => [ORIGINAL REDIRECT]
    [state] => [355 byte state value]
    [jetpack-code] =>
    [jetpack-user-id] => 0
    [action] => oauth2-login
)
```

In addition to this we have a huge 4282 byte cookie which includes:

`jetpack_sso_original_request` - a 1964 byte URL (the one in stage 3)
`jetpack_sso_redirect_to` - a 1437 byte URL (the `REDIRECT_TO` in stage 4)

In addition the referer is sent - a 1801 byte URL

The POST data contains:

`jetpack_json_api_original_query` - 1800 URL (stage 3)
`redirect_to` - 1327 byte URL (the `REDIRECT_TO` in stage 4)

At this point my default Nginx spits out a 502 error with `upstream sent too big header while reading response header from upstream` in the error log.

To summarise the request that breaks the server is:

- 1801 bytes of URL
- 4282 bytes of cookie
- 1801 bytes of referer

I don’t think it’s specifically a URL length problem, but that the total size of the HTTP header exceeds the default supported by Nginx.

To resolve the problem you can increase the size:

```
fastcgi_buffers 16 16k;
fastcgi_buffer_size 32k;
```

Note that Nginx defaults to a [4k buffer](https://nginx.org/en/docs/http/ngx_http_fastcgi_module.html#fastcgi_buffers). If my understanding of the problem is correct we've exceeded that with a request that's more than 7k.

Some bits of data seem to be repeated in several places, both in the URL (which is URL encoded, making it longer), as well as in cookies and POST data.

**Comment by chaselivingston** [comment]
Also reported in 3140666-t.

**Comment by jeherve** [comment]
Closing as duplicate of #3546. I believe this is the same issue.
